### PR TITLE
core: Add fault_exception_recovery

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds a way to mark a thread as non-essential and allow it to be marked as crashed when it triggers an fault exception on the MCU. This way non-critical threads can fail gracefully without affecting the rest of the system by triggering a reboot.

`core/thread` is extended with the logic and state to initialize a thread as `THREAD_CONFIG_KILL_ON_CRASH`, where the thread will be put into the `STATUS_CRASHED` state if it triggers an exception. Afterwards the scheduler continues as usual and the system will not panic.

the fault_exception_recovery is added to signal that a CPU implements the necessary logic to catch crashed threads. This PR contains an implementation for the cortex-m platform, covering all current cortex-m CPUs.

This module does not protect against lockups where a thread crashes after locking a mutex or crashes inside a critical section.

### Testing procedure

A test application is supplied to test and demo the functionality.

### Issues/PRs references

None

### CODE REVIEW

File: core/include/sched.h [hunk 1]
```diff
@@ -148,6 +148,7 @@ typedef struct _thread thread_t;
  */
 typedef enum {
     STATUS_STOPPED,                 /**< has terminated                           */
+    STATUS_CRASHED,                 /**< thread has crashed with a fault          */
```
Comment (benpicco) [inline core/include/sched.h:151]: You will also need to patch OpenOCD if you modify this

Comment (bergzand) [reply core/include/sched.h:151]: Yup, I'm aware. Do you happen to know how they accept patches?

Comment (benpicco) [reply core/include/sched.h:151]: They use [Gerrit](http://openocd.zylin.com/#/q/status:open) - there is a [guide](http://openocd.org/doc-release/doxygen/patchguide.html).

However, you would then have to determine which RIOT version you are on at run-time.
A lot of the hassle could be avoided if you appended the new state at the end of the list.

Comment (bergzand) [reply core/include/sched.h:151]: >  A lot of the hassle could be avoided if you appended the new state at the end of the list.

That would require rewriting the [`sched_set_status` function](https://github.com/RIOT-OS/RIOT/blob/master/core/sched.c#L226)

Comment (maribu) [reply core/include/sched.h:151]: You could still make it in time for the next OpenOCD release :-)

**Comment by fjmolinas** [comment]
@bergzand what is the status on this one? Do you plan on providing an openocd patch before moving forward?

**Pull request closed without merging**